sbtc/src/leb128.rs

```diff
@@ -0,0 +1,556 @@
+        let val3 = cursor.read_leb128().unwrap();
+        assert_eq!(val3, 255);
+        assert_eq!(cursor.position(), 5);
+
+        // Attempt read at end (should error)
+        match cursor.read_leb128() {
+            Err(Error::IndexOutOfBounds) => {} // Expected
+            other => panic!("expected IndexOutOfBounds error, got {:?}", other),
+        }
+
+        // Position should remain unchanged after error
+        assert_eq!(cursor.position(), 5);
+    }
+
+    /// Tests read operations with extremely large position values
+    #[test]
+    fn test_large_cursor_positions() {
+        // Test with position that requires careful handling to avoid overflow
+        // We'll create a slice with one valid element at the end
+        let buffer = [0x42]; // Single byte value
+        let mut cursor = Cursor::new(&buffer[..]);
+
+        // Test close to u32::MAX on 32-bit platforms and larger on 64-bit
+        let large_but_valid_pos = (usize::MAX / 2) as u64;
+
+        // This position exceeds buffer length, should return IndexOutOfBounds
+        cursor.set_position(large_but_valid_pos);
+        match cursor.read_leb128() {
+            Err(Error::IndexOutOfBounds) => {} // Expected
+            other => panic!("Expected IndexOutOfBounds error, got {:?}", other),
+        }
+    }
+
+    #[test_case(0 => Ok(127); "zero")]
+    #[test_case(1 => Err(Error::IndexOutOfBounds); "just past end")]
+    #[test_case(u64::MAX - 10 => Err(Error::IndexOutOfBounds); "potential overflow")]
+    fn test_safe_cursor_bounds_handling(pos: u64) -> Result<u64, Error> {
+        let buffer = [0x7F]; // Single value (127)
+        let mut cursor = Cursor::new(&buffer[..]);
+
+        cursor.set_position(pos);
+        cursor.read_leb128()
+    }
+}
+
+#[cfg(test)]
+mod proptests {
+    use super::*;
+    use proptest::prelude::*;
+
+    // Constants for proptest strategies
+    const MAX_PROPTEST_ITERATIONS: u32 = 10000;
+
+    // Value range constants for stratified testing. We use separate tests
+    // for these ranges as proptest uses pseudo-random sampling and otherwise
+    // may not get enough coverage for the different value ranges.
+    const VALUE_MAX_1_BYTE: u64 = (1 << 7) - 1; // 127
+    const VALUE_MAX_2_BYTES: u64 = (1 << 14) - 1; // 16,383
+    const VALUE_MAX_3_BYTES: u64 = (1 << 21) - 1; // 2,097,151
+    const VALUE_MAX_4_BYTES: u64 = (1 << 28) - 1; // 268,435,455
+    const VALUE_MAX_5_BYTES: u64 = (1 << 35) - 1; // 34,359,738,367
+    const VALUE_MAX_6_BYTES: u64 = (1 << 42) - 1; // 4,398,046,511,103
+    const VALUE_MAX_7_BYTES: u64 = (1 << 49) - 1; // 562,949,953,421,311
+    const VALUE_MAX_8_BYTES: u64 = (1 << 56) - 1; // 72,057,594,037,927,935
+    const VALUE_MAX_9_BYTES: u64 = (1 << 63) - 1; // 9,223,372,036,854,775,807
+    const VALUE_MAX_10_BYTES: u64 = u64::MAX; // 18,446,744,073,709,551,615
+
+    /// Number of payload bits per LEB128 byte (excluding continuation bit)
+    const BITS_PER_BYTE: u32 = 7;
```

Comment:
nit: why not reusing the super one?

Reply:
Well, I didn't really trust it 🤔 Nah, probably had this one first, removed in [9e11140](https://github.com/stacks-network/sbtc/pull/1523/commits/9e111402f9cc91a0c87cb6bd58d9699950fba1b0)